sound/soc/sof/intel/hda-loader.c

```diff
@@ -335,6 +335,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+static int hda_dsp_boot_imr(struct snd_sof_dev *sdev)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	const struct sof_intel_dsp_desc *chip = hda->desc;
+	int ret, i;
+
+	/* init for booting wait */
+	init_waitqueue_head(&sdev->boot_wait);
+
+	/* power up & unstall/run core 0 to run the firmware */
+	ret = hda_dsp_enable_core(sdev, BIT(0));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: dsp core start failed %d\n", ret);
+		return -EIO;
+	}
+
+	/* DSP is powered up, set all SSPs to slave mode */
+	for (i = 0; i < chip->ssp_count; i++) {
+		snd_sof_dsp_update_bits_unlocked(sdev, HDA_DSP_BAR,
+						 chip->ssp_base_offset
+						 + i * SSP_DEV_MEM_SIZE
+						 + SSP_SSC1_OFFSET,
+						 SSP_SET_SLAVE,
+						 SSP_SET_SLAVE);
+	}
+
+	/* enable IPC interrupts */
+	hda_dsp_ipc_int_enable(sdev);
+
+	/*
+	 * When a SoundWire link is in clock stop state, a Slave
+	 * device may trigger in-band wakes for events such as jack
+	 * insertion or acoustic event detection. This event will lead
+	 * to a WAKEEN interrupt, handled by the PCI device and routed
+	 * to PME if the PCI device is in D3. The resume function in
+	 * audio PCI driver will be invoked by ACPI for PME event and
+	 * initialize the device and process WAKEEN interrupt.
+	 *
+	 * The WAKEEN interrupt should be processed ASAP to prevent an
+	 * interrupt flood, otherwise other interrupts, such IPC,
+	 * cannot work normally.  The WAKEEN is handled after the ROM
+	 * is initialized successfully, which ensures power rails are
+	 * enabled before accessing the SoundWire SHIM registers
+	 */
+	if (!sdev->first_boot)
```

Comment:
On a similar note, when is the firmware preserved in IMR?
D3/S0 aka pm_runtime_suspend or RTD3?
D3/S3 aka system suspend/hibernate?

Reply:
> On a similar note, when is the firmware preserved in IMR?
> D3/S0 aka pm_runtime_suspend or RTD3?
> D3/S3 aka system suspend/hibernate?

Per my understanding, it is preserved in case the DDR is with power, so yes for both D3/S0 and D3/S3. 

I am not quite sure about the S4 (Suspend-to-Disk) case though, what do you think about that?

Reply:
That's for you to find out @keyonjie, and it goes back to my question on 'first-boot'. It's rather important to know when the IMR is preserved and when it's not.

Reply:
> That's for you to find out @keyonjie, and it goes back to my question on 'first-boot'. It's rather important to know when the IMR is preserved and when it's not.

Since I can't find a platform that support S4 hibernation at my end, will try this later.